plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/TokenAccount.hs

```diff
@@ -0,0 +1,138 @@
+import qualified Ledger.Value                                      as Value
+
+import qualified Language.PlutusTx.Coordination.Contracts.Currency as Currency
+
+newtype AccountOwner = AccountOwner { unAccountOwner :: (CurrencySymbol, TokenName) }
+    deriving newtype Eq
+
+data TokenAccount
+
+instance ScriptType TokenAccount where
+    type RedeemerType TokenAccount = ()
+    type DataType TokenAccount = AccountOwner
+
+{-# INLINEABLE accountToken #-}
+accountToken :: AccountOwner -> Value
+accountToken (AccountOwner (symbol, name)) = Value.singleton symbol name 1
+
+{-# INLINEABLE validate #-}
+validate :: AccountOwner -> () -> V.PendingTx -> Bool
+validate owner _ ptx = V.valueSpent ptx `Value.geq` accountToken owner
+
+scriptInstance :: Scripts.ScriptInstance TokenAccount
+scriptInstance =
+    Scripts.Validator @TokenAccount $$(PlutusTx.compile [|| validate ||]) $$(PlutusTx.compile [|| wrap ||])
+    where
+    wrap = Scripts.wrapValidator @AccountOwner @()
+
+address :: Address
+address = Scripts.scriptAddress scriptInstance
+
+-- | Pay some money to the given token
+pay :: (AsContractError e, HasWriteTx s) => AccountOwner -> Value -> Contract s e TxId
+pay owner vl =
+    let ds = Ledger.DataScript (PlutusTx.toData owner)
+    in writeTxSuccess (payToScript vl address ds)
+
+ownsTxOut :: AccountOwner -> TxOut -> Bool
+ownsTxOut owner txout =
+    Just owner == (Ledger.txOutData txout >>= PlutusTx.fromData . Ledger.getDataScript)
+
+-- | Create a transaction that spends all outputs belonging to the 'AccountOwner'.
+redeemTx
+    :: ( HasUtxoAt s )
+    => AccountOwner
+    -> PubKey
+    -> Contract s e UnbalancedTx
+redeemTx owner pk = do
+    utxos <- utxoAt (Scripts.scriptAddress scriptInstance)
+    let filterByOwner _ = ownsTxOut owner
+        tx = TypedTx.collectFromScriptFilter filterByOwner utxos scriptInstance ()
+    -- TODO. Replace 'PubKey' with a more general 'Address' type of output?
+    --       Or perhaps add a field 'requiredTokens' to 'UnbalancedTx' and let the
+    --       balancing mechanism take care of providing the token.
+    pure $ tx & outputs .~ [pubKeyTxOut (accountToken owner) pk]
+
+-- | Empty the account by spending all outputs belonging to the 'AccountOwner'.
+redeem
+  :: ( AsContractError e
+     , HasWriteTx s
+     , HasUtxoAt s
+     )
+  => PubKey 
+  -- ^ Where the token should go after the transaction
+  -> AccountOwner
+  -- ^ Account owner token
+  -> Contract s e TxId
+redeem pk owner = redeemTx owner pk >>= writeTxSuccess
+
+-- | @balance owners@ returns the value of all unspent outputs that can be unlocked
+--   with @accountToken owners@
+balance
+    :: ( HasUtxoAt s )
+    => AccountOwner
+    -> Contract s e Value
+balance owner = do
+    utxos <- utxoAt (Scripts.scriptAddress scriptInstance)
+    let inner =
+            foldMap (view Ledger.outValue)
+            $ filter (ownsTxOut owner)
+            $ fmap snd
+            $ Map.toList
```

Comment:
Nit: you can use `Map.filter` and then `foldMap` will work on the map values so you can drop the `Map.toList` and `fmap snd`.